WalletWasabi/Blockchain/Analysis/CoinjoinAnalyzer.cs

```diff
@@ -0,0 +1,55 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using WalletWasabi.Blockchain.Transactions;
+
+namespace WalletWasabi.Blockchain.Analysis;
+
+public class CoinjoinAnalyzer
+{
+	private Dictionary<SmartCoin, decimal> cachedInputSanctions = new();
+
+	public CoinjoinAnalyzer(SmartTransaction analyzedTransaction)
+	{
+		AnalyzedTransaction = analyzedTransaction;
+	}
+
+	public SmartTransaction AnalyzedTransaction { get; }
+
+	public decimal ComputeInputSanction(SmartCoin transactionInput)
+	{
+		HashSet<OutPoint> analyzedTransactionPrevOuts = AnalyzedTransaction.Transaction.Inputs.Select(input => input.PrevOut).ToHashSet();
+
+		decimal ComputeInputSanctionHelper(SmartCoin transactionOutput)
+		{
+			if (cachedInputSanctions.ContainsKey(transactionOutput))
+			{
+				return cachedInputSanctions[transactionOutput];
+			}
+
+			SmartTransaction transaction = transactionOutput.Transaction;
+			decimal sanction = CoinjoinAnalyzer.ComputeAnonymityContribution(transactionOutput, analyzedTransactionPrevOuts);
+			sanction += transaction.WalletInputs.Select(ComputeInputSanctionHelper).Sum();
+			cachedInputSanctions[transactionOutput] = sanction;
+			return sanction;
+		}
+
+		return ComputeInputSanctionHelper(transactionInput);
+	}
+
```

Comment:
Let's add some documentation: Computes how much the foreign outputs of AnalyzedTransaction contribute to the anonymity of our transactionOutput. Sometimes we are only interested in how much a certain subset of foreign outputs contributed. This subset can be specified in relevantOutpoints, otherwise all outputs are considered relevant.